stan/math/rev/functor/cvodes_integrator.hpp

```diff
@@ -75,116 +183,174 @@ class cvodes_integrator {
+      : f_(f),
+        y0_(y0.unaryExpr([](const T_y0& val) { return T_y0_t0(val); })),
+        t0_(t0),
+        ts_(ts),
+        args_tuple_(args...),
+        value_of_args_tuple_(value_of(args)...),
+        N_(y0.size()),
+        msgs_(msgs),
+        relative_tolerance_(relative_tolerance),
+        absolute_tolerance_(absolute_tolerance),
+        include_sensitivities_in_errors_(include_sensitivities_in_errors),
+        max_num_steps_(max_num_steps),
+        y0_vars_(count_vars(y0_)),
+        args_vars_(count_vars(args...)),
+        coupled_ode_(f, y0_, msgs, args...),
+        coupled_state_(coupled_ode_.initial_state()) {
+    const char* fun = "cvodes_integrator::integrate";
+
+    check_finite(fun, "initial state", y0_);
+    check_finite(fun, "initial time", t0_);
+    check_finite(fun, "times", ts_);
+
+    // Code from: https://stackoverflow.com/a/17340003 . Should probably do
+    // something better
+    apply(
+        [&](auto&&... args) {
+          std::vector<int> unused_temp{
+              0, (check_finite(fun, "ode parameters and data", args), 0)...};
+        },
+        args_tuple_);
+
+    check_nonzero_size(fun, "times", ts_);
+    check_nonzero_size(fun, "initial state", y0_);
+    check_sorted(fun, "times", ts_);
+    check_less(fun, "initial time", t0_, ts_[0]);
+    check_positive_finite(fun, "relative_tolerance", relative_tolerance_);
+    check_positive_finite(fun, "absolute_tolerance", absolute_tolerance_);
+    check_positive(fun, "max_num_steps", max_num_steps_);
+
+    nv_state_ = N_VMake_Serial(N_, &coupled_state_[0]);
+    nv_state_sens_ = nullptr;
+    A_ = SUNDenseMatrix(N_, N_);
+    LS_ = SUNDenseLinearSolver(nv_state_, A_);
+
+    if (y0_vars_ + args_vars_ > 0) {
+      nv_state_sens_
+          = N_VCloneVectorArrayEmpty_Serial(y0_vars_ + args_vars_, nv_state_);
+      for (std::size_t i = 0; i < y0_vars_ + args_vars_; i++) {
+        NV_DATA_S(nv_state_sens_[i]) = &coupled_state_[N_] + i * N_;
+      }
     }
-    if (max_num_steps <= 0) {
-      invalid_argument(function_name, "max_num_steps,", max_num_steps, "",
-                       ", must be greater than 0");
+  }
+
+  ~cvodes_integrator() {
+    SUNLinSolFree(LS_);
+    SUNMatDestroy(A_);
+    N_VDestroy_Serial(nv_state_);
+    if (y0_vars_ + args_vars_ > 0) {
+      N_VDestroyVectorArray_Serial(nv_state_sens_, y0_vars_ + args_vars_);
     }
+  }
 
-    const size_t N = y0.size();
-    const size_t M = theta.size();
-    const size_t S = (initial_var::value ? N : 0) + (param_var::value ? M : 0);
+  /**
+   * Solve the ODE initial value problem y' = f(t, y), y(t0) = y0 at a set of
+   * times, { t1, t2, t3, ... } using the stiff backward differentiation formula
+   * (BDF) solver in CVODES.
+   *
+   * @return std::vector of Eigen::Matrix of the states of the ODE, one for each
+   *   solution time (excluding the initial state)
+   */
+  std::vector<Eigen::Matrix<T_Return, Eigen::Dynamic, 1>> operator()() {
```

Comment:
for what is this operator needed?

Reply:
I think the two signatures thing was left over from the tolerances experiments. Removed.